File: medium/3751-total-waviness-of-numbers-in-range-i/solution.java
/**
 * 3751. Total Waviness of Numbers in Range I
 * Time: O(R * D)
 * Space: O(D)
 */
class Solution {
    public int totalWaviness(int num1, int num2) {
        int total = 0;

        for (int num = num1; num <= num2; num++) {
            total += waviness(num);
        }

        return total;
    }

    private int waviness(int num) {
        String digits = Integer.toString(num);
        int count = 0;

        for (int index = 1; index < digits.length() - 1; index++) {
            char previousDigit = digits.charAt(index - 1);
            char currentDigit = digits.charAt(index);
            char nextDigit = digits.charAt(index + 1);

            if (
                (previousDigit < currentDigit && currentDigit > nextDigit)
                    || (previousDigit > currentDigit && currentDigit < nextDigit)
            ) {
                count++;
            }
        }

        return count;
    }
}

File: medium/3751-total-waviness-of-numbers-in-range-i/solution.py
class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:
        """
        3751. Total Waviness of Numbers in Range I
        Time: O(R * D)
        Space: O(D)
        """
        def waviness(num: int) -> int:
            digits = str(num)
            count = 0

            for index in range(1, len(digits) - 1):
                previous_digit = digits[index - 1]
                current_digit = digits[index]
                next_digit = digits[index + 1]

                if (
                    previous_digit < current_digit > next_digit
                    or previous_digit > current_digit < next_digit
                ):
                    count += 1

            return count

        return sum(waviness(num) for num in range(num1, num2 + 1))
